src/storage-client/src/controller.rs

```diff
@@ -886,25 +890,57 @@ where
                 .await
                 .unwrap();
 
-            let (write, mut read) = persist_client
-                .open(metadata.data_shard)
+            let write = persist_client
+                .open_writer(metadata.data_shard)
                 .await
                 .expect("invalid persist usage");
 
-            // Advance the collection's `since` as requested.
-            if let Some(since) = &description.since {
-                read.downgrade_since(since).await;
-            }
+            // Construct the handle in a separate block to ensure all error paths are diverging
+            let since_handle = {
+                let mut handle = persist_client
+                    .open_critical_since(
+                        metadata.data_shard,
+                        PersistClient::CONTROLLER_CRITICAL_SINCE,
+                    )
+                    .await
+                    .expect("invalid persist usage");
+
+                let since = description
+                    .since
+                    .clone()
+                    .unwrap_or_else(|| handle.since().clone());
+
+                // We should only continue if we can fence out any other processes
+                let our_epoch = self.state.envd_epoch;
+                if handle.opaque() <= &our_epoch {
+                    let cur = *handle.opaque();
+                    let result = handle
+                        .maybe_compare_and_downgrade_since(&cur, (&our_epoch, &since))
```

Comment:
we want to switch this to the non "maybe_" version, right?

Reply:
That's right